ZAPD/ZCKeyFrame.cpp

```diff
@@ -0,0 +1,302 @@
+	if (limbType == ZKeyframeSkelType::Error)
+		HANDLE_ERROR_RESOURCE(
+			WarningType::InvalidXML, parent, this, rawDataIndex, "Invalid limb type",
+			StringHelper::Sprintf("Invalid limb type. Was expecting 'Flex' or 'Normal'. Got %s.",
+		                          limbTypeStr.c_str()));
+}
+void ZKeyFrameLimbList::ParseXML(tinyxml2::XMLElement* reader)
+{
+	ZResource::ParseXML(reader);
+
+	std::string limbTypeStr = registeredAttributes.at("LimbType").value;
+	std::string numLimbStr = registeredAttributes.at("LimbCount").value;
+
+	limbType = ParseLimbTypeStr(limbTypeStr);
+
+	if (limbType == ZKeyframeSkelType::Error)
+		HANDLE_ERROR_RESOURCE(
+			WarningType::InvalidXML, parent, this, rawDataIndex, "Invalid limb type",
+			StringHelper::Sprintf("Invalid limb type. Was expecting 'Flex' or 'Normal'. Got %s.",
+		                          limbTypeStr.c_str()));
+
+	numLimbs = (uint8_t)StringHelper::StrToL(numLimbStr);
+}
+
+void ZKeyFrameSkel::ParseRawData()
+{
+	ZResource::ParseRawData();
+
+	const auto& rawData = parent->GetRawData();
+	limbCount = BitConverter::ToUInt8BE(rawData, rawDataIndex + 0);
+	dListCount = BitConverter::ToUInt8BE(rawData, rawDataIndex + 1);
+	limbsPtr = BitConverter::ToUInt32BE(rawData, rawDataIndex + 4);
+
+	limbList = std::make_unique<ZKeyFrameLimbList>(parent, limbCount, limbType);
+	limbList->SetRawDataIndex(GETSEGOFFSET(limbsPtr));
+	limbList->ParseRawData();
+}
+
+void ZKeyFrameSkel::DeclareReferences(const std::string& prefix)
+{
+	std::string defaultPrefix = name;
+	std::string declaration;
+
+	if (defaultPrefix == "")
+		defaultPrefix = prefix;
+
+	ZResource::DeclareReferences(defaultPrefix);
+	declaration += limbList->GetBodySourceCode();
+	parent->AddDeclarationArray(
+		GETSEGOFFSET(limbsPtr), DeclarationAlignment::Align4, limbList->GetRawDataSize(),
+		limbList->GetSourceTypeName(),
+		StringHelper::Sprintf("%s_KeyFrameLimbs_%06X", prefix.c_str(), rawDataIndex),
+		limbList->limbs.size(), declaration);
+}
+
+std::string ZKeyFrameSkel::GetBodySourceCode() const
+{
+	std::string limbStr;
+
+	if (limbType == ZKeyframeSkelType::Normal)
+		Globals::Instance->GetSegmentedPtrName(limbsPtr, parent, "KeyFrameStandardLimb", limbStr);
+	else
+		Globals::Instance->GetSegmentedPtrName(limbsPtr, parent, "KeyFrameFlexLimb", limbStr);
+
+	return StringHelper::Sprintf("\n\t0x%02X, 0x%02X, %s\n", limbCount, dListCount,
+	                             limbStr.c_str());
+}
+
+size_t ZKeyFrameSkel::GetRawDataSize() const
+{
+	return 0x8;
+}
+
+std::string ZKeyFrameSkel::GetSourceTypeName() const
+{
+	return "KeyFrameSkeleton";
+}
+
+ZResourceType ZKeyFrameSkel::GetResourceType() const
+{
+	return ZResourceType::KeyFrameSkel;
+}
+
+size_t ZKeyFrameStandardLimb::GetRawDataSize() const
+{
+	return 0xC;
+}
+
+size_t ZKeyFrameFlexLimb::GetRawDataSize() const
+{
+	return 0x8;
+}
+
+size_t ZKeyFrameLimbList::GetRawDataSize() const
+{
+	size_t limbSize;
+	if (limbType == ZKeyframeSkelType::Flex)
+		limbSize = 8;
+	else
+		limbSize = 0xC;
```

Comment:
It's weird having decimal for one size and hex for another. I'd say do one or the other.